samples/JwtBearerSample/wwwroot/App/Scripts/adal.js

```diff
@@ -0,0 +1,1060 @@
+    };
+
+    return crackedToken;
+};
+
+AuthenticationContext.prototype._convertUrlSafeToRegularBase64EncodedString = function (str) {
+    return str.replace('-', '+').replace('_', '/');
+};
+
+AuthenticationContext.prototype._serialize = function (responseType, obj, resource) {
+    var str = [];
+    if (obj !== null) {
+        str.push('?response_type=' + responseType);
+        str.push('client_id=' + encodeURIComponent(obj.clientId));
+        if (resource) {
+            str.push('resource=' + encodeURIComponent(resource));
+        }
+
+        str.push('redirect_uri=' + encodeURIComponent(obj.redirectUri));
+        str.push('state=' + encodeURIComponent(obj.state));
+
+        if (obj.hasOwnProperty('slice')) {
+            str.push('slice=' + encodeURIComponent(obj.slice));
+        }
+
+        if (obj.hasOwnProperty('extraQueryParameter')) {
+            str.push(obj.extraQueryParameter);
+        }
+    }
+
+    return str.join('&');
+};
+
+AuthenticationContext.prototype._deserialize = function (query) {
+    var match,
+        pl = /\+/g,  // Regex for replacing addition symbol with a space
+        search = /([^&=]+)=?([^&]*)/g,
+        decode = function (s) { return decodeURIComponent(s.replace(pl, ' ')); },
+        obj = {};
+    match = search.exec(query);
+    while (match) {
+        obj[decode(match[1])] = decode(match[2]);
+        match = search.exec(query);
+    }
+
+    return obj;
+};
+
+/* jshint ignore:start */
+AuthenticationContext.prototype._guid = function () {
+    // RFC4122: The version 4 UUID is meant for generating UUIDs from truly-random or
+    // pseudo-random numbers.
+    // The algorithm is as follows:
+    //     Set the two most significant bits (bits 6 and 7) of the
+    //        clock_seq_hi_and_reserved to zero and one, respectively.
+    //     Set the four most significant bits (bits 12 through 15) of the
+    //        time_hi_and_version field to the 4-bit version number from
+    //        Section 4.1.3. Version4 
+    //     Set all the other bits to randomly (or pseudo-randomly) chosen
+    //     values.
+    // UUID                   = time-low "-" time-mid "-"time-high-and-version "-"clock-seq-reserved and low(2hexOctet)"-" node
+    // time-low               = 4hexOctet
+    // time-mid               = 2hexOctet
+    // time-high-and-version  = 2hexOctet
+    // clock-seq-and-reserved = hexOctet: 
+    // clock-seq-low          = hexOctet
+    // node                   = 6hexOctet
+    // Format: xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx
+    // y could be 1000, 1001, 1010, 1011 since most significant two bits needs to be 10
+    // y values are 8, 9, A, B
+    var guidHolder = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx';
+    var hex = '0123456789abcdef';
+    var r = 0;
+    var guidResponse = "";
+    for (var i = 0; i < 36; i++) {
+        if (guidHolder[i] !== '-' && guidHolder[i] !== '4') {
+            // each x and y needs to be random
+            r = Math.random() * 16 | 0;
```

Comment:
Math.random (at least in chrome) might not be the best thing to use for generating globally unique identifiers. Here's a recent blog post about it:

https://medium.com/@betable/tifu-by-using-math-random-f1c308c4fd9d#.ey40psqsl

In the blog post, the generated a "unique" id with supposedly 132 bits of entropy, but saw real world collisions because of Chromes PRNG.

Can we generate the GUID on the server instead of using Math.random?